### PR TITLE
[ISSUE #855] Add the Test Util And ORC Writer for Spark

### PR DESCRIPTION
**Pull request by shawnding**
Add the Test Util And ORC file writer.
Lack of the metrics information and file length check.

### CODE REVIEW

File: orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java [hunk 1]
```diff
@@ -96,8 +96,8 @@ public Metrics metrics() {
 
   @Override
   public long length() {
-    Preconditions.checkState(isClosed,
-        "Cannot return length while appending to an open file.");
+//    Preconditions.checkState(isClosed,
```
Comment (chenjunjiedada) [inline orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java]: Why comment out this?

File: spark/src/test/java/org/apache/iceberg/spark/source/NestedRecord.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+package org.apache.iceberg.spark.source;
```
Comment (chenjunjiedada) [inline spark/src/test/java/org/apache/iceberg/spark/source/NestedRecord.java]: Please add apache license

File: spark/src/test/java/org/apache/iceberg/spark/source/SimpleRecord.java [hunk 3]
```diff
@@ -20,6 +20,9 @@
 package org.apache.iceberg.spark.source;
 
 import com.google.common.base.Objects;
+import com.oracle.tools.packager.mac.MacAppBundler;
```
Comment (chenjunjiedada) [inline spark/src/test/java/org/apache/iceberg/spark/source/SimpleRecord.java]: Please remove these unrelated changes.

File: spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
+package org.apache.iceberg.spark.source;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.*;
```
Comment (chenjunjiedada) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java]: The Iceberg code style doesn't allow using the wildcard import. You might want to change your IDE setting to disable the import optimization.

File: spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.iceberg.hadoop.HadoopTables;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Encoders;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.junit.*;
```
Comment (chenjunjiedada) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java]: ditto.

File: spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
```
Comment (chenjunjiedada) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java]: Please sort the import alphabetically.

File: spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        .option("write-format", "orc")
+        .mode("append")
+        .save(location.toString());
+
+    table.refresh();
+
+    Dataset<Row> result = spark.read()
+        .format("iceberg")
+        .load(location.toString());
+
```
Comment (chenjunjiedada) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java]: unnecessary blank line.

File: spark/src/main/java/org/apache/iceberg/spark/source/Writer.java [hunk 8]
```diff
@@ -308,6 +310,13 @@ public String toString() {
                   .schema(dsSchema)
                   .overwrite()
                   .build();
+            case ORC:
```
Comment (chenjunjiedada) [inline spark/src/main/java/org/apache/iceberg/spark/source/Writer.java]: Please add a blank line before this.

File: spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
+package org.apache.iceberg.spark.source;
+
+import com.google.common.collect.ImmutableMap;
```
Comment (rdsr) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java]: @shawnding  what are your thoughts on making the class `org.apache.iceberg.spark.source.TestParquetWrite` parameterized and making it work for both Parquet and ORC. Seems like that's less work and it also tests with a lot more test cases. Thoughts? 

If you want to see some examples of parameterized tests, please see `org.apache.iceberg.data.TestLocalScan`

File: spark/src/test/java/org/apache/iceberg/spark/source/NestedRecord.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class NestedRecord {
+  private Integer id;
+  private String data;
+  private Map info;// = new HashMap<String, String>();
+
+  public NestedRecord() {
```
[the remark this answers is not in the document]
Comment (shawnding) [reply spark/src/test/java/org/apache/iceberg/spark/source/NestedRecord.java]: I think, ORC file not support Nested type, so I want to confirmed that could work correctly

File: spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java (hunk 9 above)
Comment (shawnding) [reply spark/src/test/java/org/apache/iceberg/spark/source/TestORCWrite.java]: Ok thanks Rdsr, that I will rewrite it.

**Comment by shawnding** [comment]
HI @rdsr , would you mind to take another look or give some advices on next steps？ thanks

**Comment by rdsr** [comment]
Thanks @shawnding . I'll try to review this today!

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 11]
```diff
@@ -269,7 +269,7 @@ private static TypeDescription buildOrcProjection(Integer fieldId, Type type, bo
         break;
       case MAP:
         Types.MapType map = (Types.MapType) type;
-        TypeDescription keyType = buildOrcProjection(map.keyId(), map.keyType(), true, mapping);
```
Comment (rdsr) [inline orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java]: Why was this change required?

File: spark/src/main/java/org/apache/iceberg/spark/source/Writer.java [hunk 12]
```diff
@@ -389,7 +399,9 @@ public EncryptedOutputFile newOutputFile(PartitionKey key) {
     public abstract void write(InternalRow row) throws IOException;
 
     public void writeInternal(InternalRow row)  throws IOException {
-      if (currentRows % ROWS_DIVISOR == 0 && currentAppender.length() >= targetFileSize) {
+      //TODO: ORC file now not support target file size
```
Comment (rdsr) [inline spark/src/main/java/org/apache/iceberg/spark/source/Writer.java]: What exception you see if when running with ORC? I would have thought that it should work fine.
Also I think it is  disabled for Avro also now.

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcReader.java [hunk 13]
```diff
@@ -49,7 +49,7 @@ protected void writeAndValidate(Schema schema) throws IOException {
     }
 
     try (CloseableIterable<InternalRow> reader = ORC.read(Files.localInput(testFile))
-        .schema(schema)
+        .project(schema)
```
Comment (rdsr) [inline spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcReader.java]: nit: would it be easier to address this change as a separate PR. Seems un-related to current work

File: spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataWrite.java [hunk 14]
```diff
@@ -0,0 +1,457 @@
...
+    Assert.assertEquals("Result rows should match", expected, actual);
+
+    List<DataFile> files = Lists.newArrayList();
+    for (ManifestFile manifest : table.currentSnapshot().manifests()) {
+      for (DataFile file : ManifestReader.read(manifest, table.io())) {
+        files.add(file);
+      }
+    }
+    // TODO: ORC file now not support target file size
+    if (format.equals(FileFormat.PARQUET)) {
```
Comment (rdsr) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataWrite.java]: Does this not work for Avro as well?

File: spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataWrite.java [hunk 15]
```diff
@@ -0,0 +1,457 @@
+/*
```
Comment (rdsr) [inline spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataWrite.java]: Does this file replace `org.apache.iceberg.spark.source.TestParquetWrite` or do u think both are required?

Comment (shawnding) [reply spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataWrite.java]: yes , I Think this file replace `org.apache.iceberg.spark.source.TestParquetWrite`

File: spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataWrite.java (hunk 14 above)
Comment (shawnding) [reply spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataWrite.java]: Yes now `target-file-size-byte` not work on Avro or Orc file.

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java (hunk 11 above)
Comment (shawnding) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java]: While a `MapType ` column is optional if insert a  empty map like `null`,  the `key` and `value`  both are `null`

File: spark/src/main/java/org/apache/iceberg/spark/source/Writer.java (hunk 12 above)
Comment (shawnding) [reply spark/src/main/java/org/apache/iceberg/spark/source/Writer.java]: Yes in Avro file is work,  cannot get the file length before closed the  ORC file, I have fix it.

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcReader.java (hunk 13 above)
Comment (shawnding) [reply spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcReader.java]: OK I split it to another PR

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java (hunk 11 above)
Comment (rdblue) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java]: Map keys are always required. Does this make the map key optional?

**Comment by rdblue** [comment]
This looks good to me, except for two things:
1. Map keys are always required if the map is projected, so I'm not sure the map change is correct.
2. I think it would make more sense to add nested tests later. The majority of these changes are test changes to add a map. I don't think the amount of change here is worth such a narrow set of cases -- one string map with the same key/values each time.

**Comment by rdblue** [comment]
One more thing that I thought of: if this is adding the ability to write ORC files, then we should make sure all of the parameterized Spark tests that currently test Parquet and Avro are also applied to ORC.

**Comment by shawnding** [comment]
Thanks for your suggestion @rdblue . I will append same complicated nested type to this test。And now if  all of the parameter test applied to ORC file should after https://github.com/apache/incubator-iceberg/pull/199 have merged

**Comment by rdblue** [comment]
> I will append same complicated nested type to this test

Sorry for the miscommunication: I'm asking you to remove the more complicated nested type from this PR. Let's get the existing tests running on ORC for this PR and then we can add more test cases in a separate one.

**Comment by shawnding** [comment]
Thanks @rdblue , I was  removed the  nested type.would you mind to take another look or give some advices on next steps？

**Comment by rdblue** [comment]
Looks good to me. Thanks for adding this, @shawnding!
